dhis2/dhis2-core
# feat: job to disable inactive users [DHIS2-6720]

### Summary
Adds a new job type to the scheduler.
The job allows to mark users disabled that have not been active for a configurable number of months.

Inactivity is evaluated based on their `lastLogin` timestamp and the current time.
A DB update statement is used to change the `disabled` flag column for matching users.

Users that have no lastLogin set (null) are not disabled. This might very well be a theoretical case that does not exist in practice.

The service layer will not perform the operation if a date is given that is not at least 1 month in the past to prevent accidentally disabling all users by passing in a data in the future or the very recent past.

### Automatic Testing
A spring based test was added to verify the correct behaviour of the new `UserService#disableUsersInactiveSince` method.
In this context the `UserServiceTest` was cleaned by removing duplication in user setup and result set assertions using new helper methods.

### Manual Testing
* start server
* open _Scheduler_ app
* press add button
* provide a name and select job type _DISABLE_INACTIVE_USERS_
* provide a frequency and the job specific parameter _inactive months_ (number of month a user at least was inactive to be disabled)
* run the task manually

To check that the job is working as expected you either need a database with users that have not logged in for at least a month (and chose `1` for inactive months) or you need to update the DB manually before running the job to create users that have been inactive for a long while by setting column `lastLogin` in table `users` to a past date.

When the job did run it notifies about the number of users that were disabled as a consequence of the job.
With the `InMemoryNotifier` (default) this message can be found in the server logs.

Example of the notification done by the job:
```
INFO  2021-02-09T09:38:50,856 [Level: INFO, category: DISABLE_INACTIVE_USERS, time: Tue Feb 09 09:38:50 CET 2021, message: Disabled 0 users] (InMemoryNotifier.java [taskScheduler-21])
```
Note the message: `Disabled 0 users`
